### PR TITLE
Add brackets support for nested

### PR DESCRIPTION
**Pull request by nikitagromov**
Pivotal story id [#132842963](https://www.pivotaltracker.com/story/show/132842963)

### CODE REVIEW

**Comment by nikitagromov** [comment]
I've removed _nested, and create es_q param.
Because rules which were described in the story couldn't be implemented only inside nested query. 
In this case, we need to use combined query which can include nested, and other terms.
I will update docs tomorrow.

File: nefertari/es_query.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        tree = _build_tree(query_tokens)
+        return {'bool': _build_es_query(tree)}
+    return {'bool': {'must': [_parse_term(query_string)]}}
+
+
+# split query string to tokens "(", ")", "field:value", "AND", "AND NOT", "OR", "OR NOT"
+def _get_tokens(values):
+
+    tokens = []
+    brackets = ['(', ')']
```
Comment (nmenardg) [inline nefertari/es_query.py]: this could be a set:
`{'(', ')'}`

File: nefertari/es_query.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    buffer = ''
+    keywords = ['AND', 'OR']
```
Comment (nmenardg) [inline nefertari/es_query.py]: Same here, could be a set

File: nefertari/es_query.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    return tokens
+
+
+def _build_tree(tokens):
+
+    class Node:
+        def __init__(self, prev=None, next=None):
+            self.prev = prev
+            self.next = next
+            self.values = []
```
Comment (nmenardg) [inline nefertari/es_query.py:72]: `next()` is a builtin function, it's a best practice not to name arguments/variables with builtin names

File: nefertari/es_query.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+        head.values.append(token)
+    return head.parse()
+
+
+def _build_es_query(values):
+    aggregation = {}
+    operations_stack = OperationStack()
+    values_stack = []
+    keywords = ['AND', 'AND NOT', 'OR', 'OR NOT']
```
Comment (nmenardg) [inline nefertari/es_query.py]: Should be a set

File: nefertari/es_query.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+                else:
+                    for item in [value1, value2]:
+                        _attach_item(item, aggregation, operation)
+
+            values_stack.append(None)
+
+    return aggregation
+
+
+# add term to existed aggregation
```
Comment (nmenardg) [inline nefertari/es_query.py]: If this is a comment for the function, write it as a doc-string

File: nefertari/es_query.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+def _attach_item(item, aggregation, operation):
+
+    if item is None:
+        return
+
+    # init value or get existed
+    aggregation[operation] = aggregation[operation] if len(aggregation.get(operation, [])) else []
```
Comment (nmenardg) [inline nefertari/es_query.py]: Would simply `aggregation.get(operation, [])` do the job?

File: nefertari/es_query.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    if _is_nested(item):
+        _attach_nested(item, aggregation, operation)
+    elif isinstance(item, dict):
+        aggregation[operation].append(item)
+    else:
+        aggregation[operation].append(_parse_term(item))
+
+
+# parse term, on this level can be implemented rules according to range, term, match and others
+# https://www.elastic.co/guide/en/elasticsearch/reference/2.1/term-level-queries.html
```
Comment (nmenardg) [inline nefertari/es_query.py]: Docstring please

File: nefertari/es_query.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+    if '|' in value:
+        values = value.split('|')
+        return {'bool': {'should': [{'term': {field: value}} for value in values]}}
+    if value == '_missing_':
+        return {'missing': {'field': field}}
+
+    return {'term': {field: value}}
+
+
+# attach _nested to nested_document
```
Comment (nmenardg) [inline nefertari/es_query.py]: Docstring please

File: nefertari/es_query.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+def _is_nested(item):
+    if isinstance(item, str):
+        field, _ = item.split(':')
+        return '_nested' in field
+    return False
+
+
+# rules related to nested queries
+# https://www.elastic.co/guide/en/elasticsearch/guide/current/nested-query.html
```
Comment (nmenardg) [inline nefertari/es_query.py]: Docstring please

File: tests/test_es_query.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                                                                                      'path': 'assignments_nested'}}]}},
+                                                         {'bool':
+                                                              {'must': [{'term': {'value': 'true'}},
+                                                                            {'term': {'another': 'false'}},
+                                                                            {'bool': {'must_not': [{'term': {'field': 'true'}}],
+                                                                                      'must': [{'term': {'some': 'true'}}]}}]}}],
+                                              'must_not': [{'bool': {
+                                                  'should': [{'term': {'complicated': 'true'}}],
+                                                  'should_not': [{'term': {'complicated': 'false'}}]}}]}},
+                                        {'term': {'owner_id': 'someuser'}}], 'must_not': [{'term': {'completed': 'false'}}]}}
```
Comment (nmenardg) [inline tests/test_es_query.py]: Is it just github that's displaying this dictionary with a weird indentation?

**Review by nmenardg — CHANGES_REQUESTED** [review]

**Review by nmenardg — APPROVED** [review]
I guess there's a limit to how beautiful these can get XD

**Pull request merged**